provider/pkg/client/client.go

```diff
@@ -215,3 +222,142 @@ func (*clientImpl) getResourceRetryNotFoundRetrySettings() (int, *retry.Exponent
+		}
+	default:
+		// For other error codes (InvalidRequest, AccessDenied, etc.), return a
+		// generic error that won't be retried.
+		return fmt.Errorf("operation %s failed with %q: %s", pi.Operation, pi.ErrorCode, statusMessage)
+	}
+}
+
+// withRetries wraps a CloudControl operation with SDK-like retries. It
+// executes the operation and retries based on the configured Retryer,
+// imitating behavior users would see if they had performed the operation
+// directly through the SDK.
+//
+// This is necessary because our SDK config only handles retries when
+// communicating with Cloud Control, but all of CC's operations happen
+// asynchronously with their own (default) retry logic, meaning they are
+// limited to the default of 3 retries and Throttling errors aren't handled.
+func (c *clientImpl) withRetries(
+	ctx context.Context,
+	urn resource.URN,
+	operation func() (*types.ProgressEvent, error),
+) (pi *types.ProgressEvent, err error) {
+	retryer := c.retryer()
+	maxAttempts := retryer.MaxAttempts()
+
+	releaseInitialToken := retryer.GetInitialToken()
+	defer func() {
+		_ = releaseInitialToken(err)
+	}()
+
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		pi, err = operation()
+
+		// Nothing else to do if we succeeded, or if we weren't able to get a
+		// progress event. (At this point the SDK will have already retried
+		// this request for us.)
+		if err == nil || pi == nil {
+			return pi, err
+		}
+
+		// Re-hydrate the progress event into an error our Retryer can work with.
+		sdkErr := progressEventToError(pi)
+
+		if !retryer.IsErrorRetryable(sdkErr) {
+			return pi, err // Nothing else to do.
+		}
+
+		// Try to get a retry token (handles rate limiting)
+		releaseToken, tokenErr := retryer.GetRetryToken(ctx, sdkErr)
+		if tokenErr != nil {
+			if c.logger != nil {
+				_ = c.logger.LogStatus(ctx, diag.Warning, urn, fmt.Sprintf(
+					"Couldn't acquire retry token: %s", tokenErr,
+				))
+			}
+			return pi, err
+		}
+
+		// Check if we've exhausted retry attempts
+		if attempt >= maxAttempts-1 {
+			_ = releaseToken(sdkErr)
+			if c.logger != nil {
+				_ = c.logger.LogStatus(ctx, diag.Warning, urn, "Exhausted retry attempts")
+			}
+			return pi, err
+		}
+
+		// Calculate retry delay using the SDK's backoff strategy.
+		delay, delayErr := retryer.RetryDelay(attempt+1, sdkErr)
+		if delayErr != nil {
+			_ = releaseToken(sdkErr)
+			if c.logger != nil {
+				_ = c.logger.LogStatus(ctx, diag.Warning, urn, fmt.Sprintf(
+					"Couldn't calculate retry delay: %s", delayErr,
+				))
+			}
+			return pi, err
+		}
+
+		if c.logger != nil {
+			_ = c.logger.LogStatus(ctx, diag.Info, urn, fmt.Sprintf(
+				"Retrying after %q error (attempt %d/%d, waiting %s)", pi.ErrorCode, attempt+1, maxAttempts, delay,
+			))
+		}
+
+		// Wait to retry.
+		select {
+		case <-ctx.Done():
+			_ = releaseToken(sdkErr)
+			return nil, ctx.Err()
+		case <-time.After(delay):
+			_ = releaseToken(nil) // nil indicates retry will proceed.
```

Comment:
I'm sure you'd like to have `releaseToken` be in a defer block.  Line 357 seems to be the only special case, so maybe you could just set `sdkErr = nil` here and do `defer { releaseToken(sdkErr) }`.

Reply:
Yeah, normally I would prefer to defer this but we're in a loop, so the defer would only happen on function exit. The retryer uses a token bucket to throttle itself, so it's important to release the retry token as we go.